PaypalProvider: allow to make also get requests

PetrDlouhy opened the pull request:
PaypalProvider: Add also `get()` method to be able to use the provider to create also GET request to PayPal API.

radekholy24 commented:
@PetrDlouhy, I am not sure what is your use case but if you are going to use that proposed `.get(...)` (or `.http_request(..., http_method=requests.get)`) in *our* project, consider the issue described in https://github.com/jazzband/django-payments/pull/440 (i.e. that it overwrites `payment.extra_data["response"]`).

WhyNotHugo reviewed (approved):
LGTM, only one minor detail.

PetrDlouhy committed:
feat(paypal): add GET method and support requests without payment

Add get() method to complement the existing post() method, and refactor
to support making API requests that are not tied to a specific payment
instance.

Changes:
- Add http_request() helper method that accepts http_method parameter
- Refactor post() to use http_request() with requests.post
- Add new get() method that uses http_request() with requests.get
- Update http_request() to handle payment=None by skipping payment-specific
  operations (set_error_data, change_status, set_response_data)
- Update get_access_token() to handle payment=None for token caching
- Add comprehensive docstrings to both get() and post() methods

This enables making GET requests to PayPal API and allows use cases where
API calls are not associated with a payment (e.g., fetching account info,
checking subscription status).

PetrDlouhy committed:
test(paypal): add test for get() method with payment=None

Add comprehensive test to verify that get() method works correctly when
called with payment=None, which is needed for API calls not tied to a
specific payment instance.

The test verifies:
- OAuth token acquisition happens correctly
- GET request is made with proper headers
- Response data is returned correctly

PetrDlouhy committed:
fix(paypal): handle payment=None in 401 error retry and fix line length

Fix critical bug where @authorize decorator would crash on 401 errors when
payment=None, preventing proper token refresh for API calls not tied to a
payment instance.

Changes:
- Add payment=None check in @authorize decorator before accessing payment attrs
- Extract expires_in variable to fix line length linting error (line 198)
- Add test for 401 error handling with payment=None

This enables reliable use of get()/post() with payment=None for operations
like subscription management that don't have an associated Payment instance.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

WhyNotHugo commented on payments/paypal/__init__.py:
```diff
@@ -114,20 +114,21 @@ def _get_links(self, payment):
 
     @authorize
-    def post(self, payment, *args, **kwargs):
+    def http_request(self, payment, *args, http_method=None, **kwargs):
```
`method` here should be a string…

WhyNotHugo commented on payments/paypal/__init__.py:
```diff
@@ -114,20 +114,21 @@ def _get_links(self, payment):
         if "data" in kwargs:
             kwargs["data"] = json.dumps(kwargs["data"])
-        response = requests.post(*args, **kwargs)
+        response = http_method(*args, **kwargs)
```
… and we can use `requests.request(method,…` here.

WhyNotHugo commented on payments/paypal/__init__.py:
```diff
@@ -139,29 +140,66 @@ def post(self, payment, *args, **kwargs):
+            dict: JSON response data from PayPal API
+
+        Raises:
+            PaymentError: If the API returns an error status code
```
Please use standard Python format for this, not prose. E.g.:

```
:param payment: Payment instance (can be None for requests not tied to a payment)
:returns: JSON response data from PayPal API
```

Sphinx will then render these fields properly.

WhyNotHugo commented on payments/paypal/__init__.py:
```diff
@@ -139,29 +140,66 @@ def post(self, payment, *args, **kwargs):
+            self.set_response_data(payment, data)
         return data
 
+    def post(self, payment, *args, **kwargs):
```
`-> dict`.

WhyNotHugo reviewed (changes requested):
Sorry for the delay responding. Mostly minor changes required, looks good to me at a high level.

PetrDlouhy committed:
refactor(paypal): address review — string method param, Sphinx docstrings

Apply the review feedback from #439:
- http_request() takes the HTTP method as a string and dispatches via
  requests.request(method, ...) instead of receiving a requests function
  in the http_method parameter; post()/get() pass "post"/"get".
- Add a docstring to http_request() and convert the post()/get()
  docstrings to Sphinx field format (:param:/:returns:/:raises:) so the
  fields render in the docs.
- Add `-> dict` return annotations to http_request(), post() and get().
Adjust the tests for the changed mock surface: OAuth token requests
still go through requests.post, but API calls now go through
requests.request, so the tests that previously served both from a
single requests.post mock now mock the two separately (and the refund
tests assert the method argument as well).

PetrDlouhy commented:
@WhyNotHugo Thank you very much for your review. Now I fixed all your calls.

WhyNotHugo reviewed (approved):
LGTM

PetrDlouhy commented:
@WhyNotHugo Thank you very much.
